test/functional/lua/watch_spec.lua

```diff
@@ -21,30 +21,87 @@ describe('vim._watch', function()
   end)
 
   local function run(watchfunc)
-    it('detects file changes (watchfunc=' .. watchfunc .. '())', function()
+    -- Monkey-patches vim.notify_once so we can "spy" on it.
+    local function spy_notify_once()
+      exec_lua [[
+        _G.__notify_once_msgs = {}
+        vim.notify_once = (function(overridden)
+          return function(msg, level, opts)
+            table.insert(_G.__notify_once_msgs, msg)
+            return overridden(msg, level, opts)
+          end
+        end)(vim.notify_once)
+      ]]
+    end
+
+    local function last_notify_once_msg()
+      return exec_lua 'return _G.__notify_once_msgs[#_G.__notify_once_msgs]'
+    end
+
+    local function do_watch(root_dir, watchfunc_)
+      exec_lua(
+        [[
+          local root_dir, watchfunc = ...
+
+          _G.events = {}
+
+          _G.stop_watch = vim._watch[watchfunc](root_dir, {
+            debounce = 100,
+            include_pattern = vim.lpeg.P(root_dir) * vim.lpeg.P("/file") ^ -1,
+            exclude_pattern = vim.lpeg.P(root_dir .. '/file.unwatched'),
+          }, function(path, change_type)
+            table.insert(_G.events, { path = path, change_type = change_type })
+          end)
+      ]],
+        root_dir,
+        watchfunc_
+      )
+    end
+
+    it(watchfunc .. '() ignores nonexistent paths', function()
+      if watchfunc == 'inotify' then
+        skip(n.fn.executable('inotifywait') == 0, 'inotifywait not found')
+        skip(is_os('bsd'), 'inotifywait on bsd CI seems to expect path to exist?')
+      end
+
+      local msg = ('watch.%s: ENOENT: no such file or directory'):format(watchfunc)
+
+      spy_notify_once()
+      do_watch('/i am /very/funny.go', watchfunc)
+
+      if watchfunc ~= 'inotify' then -- watch.inotify() doesn't (currently) call vim.notify_once.
+        t.retry(nil, 2000, function()
+          t.eq(msg, last_notify_once_msg())
+        end)
+      end
+      eq(0, exec_lua [[return #_G.events]])
+
+      exec_lua [[_G.stop_watch()]]
+    end)
+
+    it(watchfunc .. '() detects file changes', function()
       if watchfunc == 'inotify' then
         skip(is_os('win'), 'not supported on windows')
         skip(is_os('mac'), 'flaky test on mac')
-        skip(
-          not is_ci() and n.fn.executable('inotifywait') == 0,
-          'inotify-tools not installed and not on CI'
-        )
+        skip(not is_ci() and n.fn.executable('inotifywait') == 0, 'inotifywait not found')
       end
 
+      -- Note: because this is not `elseif`, BSD is skipped for *all* cases...?
       if watchfunc == 'watch' then
         skip(is_os('mac'), 'flaky test on mac')
         skip(is_os('bsd'), 'Stopped working on bsd after 3ca967387c49c754561c3b11a574797504d40f38')
+      elseif watchfunc == 'watchdirs' and is_os('mac') then
+        -- Bump this (or fix the bug) if CI continues to fail in future versions of macos CI.
+        skip(is_ci() and vim.uv.os_uname().release == '23.6.0', 'weird failure for macOS arm 14 CI')
```

Comment:
Very strange failure only in macos 14 CI (darwin 23.6.0). I can't see how the test changes would cause this (clear() is correctly invoked for each test, and there are no logical changes...).

```
FAILED   test/functional/lua/watch_spec.lua @ 82: vim._watch watchdirs() detects file changes
test/functional/lua/watch_spec.lua:149: Expected objects to be the same.
Passed in:
(table: 0x0116023758) {
 *[1] = {
    [change_type] = 3
   *[path] = '/Users/runner/work/neovim/neovim/build/Xtest_tmpdir/nvim_KFMvPbXk9a/nvim_KFMvPbXk9a' }
  [2] = {
    [change_type] = 3
    [path] = '/Users/runner/work/neovim/neovim/build/Xtest_tmpdir/nvim_KFMvPbXk9a/file' } }
Expected:
(table: 0x010d9d6548) {
 *[1] = {
    [change_type] = 1
   *[path] = '/Users/runner/work/neovim/neovim/build/Xtest_tmpdir/nvim_KFMvPbXk9a/file' }
  [2] = {
    [change_type] = 3
    [path] = '/Users/runner/work/neovim/neovim/build/Xtest_tmpdir/nvim_KFMvPbXk9a/file' } }

stack traceback:
	test/functional/lua/watch_spec.lua:149: in function <test/functional/lua/watch_spec.lua:82>
```